### PR TITLE
Core downloads snapshot automatically - Closes #919

### LINKED ISSUE #919: Core downloads snapshot automatically
### Description

To make the start quicker, the core microservice needs to download the blockchain index database snapshot during start automatically, once a known network is detected.

### Acceptance Criteria

- Snapshot URL is defined in the config per network
- Download is started immediately
- A config option for skipping import and making a full rebuild from scratch

### PR DESCRIPTION
**Pull request by nagdahimanshu**
### What was the problem?
This PR resolves #919 

### How was it solved?
- [x] Download snapshot during the start
- [x] Add implementation to download and extract .gz file
- [x] Apply snapshot
- [x] Added configuration for snapshot
- [x] Code refactoring to avoid code duplications
- [x] Functional test cases for utility file downloadFile.js
- [x] Install MySQL client while building core service
- [x] Update config_options.md
- [x] Add relevant env variable to docker-compose file

### How was it tested?
Against `Mainnet`
- [x] Local
- [x] Linux
- [x] Docker

### CODE REVIEW

File: services/core/config.js [hunk 1]
```diff
@@ -133,4 +136,11 @@ config.queue = {
  */
 config.jobs = {};
 
+config.snapshot = {
+	url: process.env.INDEX_SNAPSHOT_URL,
+	composeFilePath: process.env.DOCKER_COMPOSE_FILEPATH,
+	serviceName: process.env.DOCKER_SERVICE_NAME,
+	enable: Boolean(String(process.env.ENABLE_APPLY_SNAPSHOT).toLowerCase() !== 'false'), // Enable by default
```
Comment (sameersubudhi) [inline services/core/config.js:143]: I think snapshot application should be disabled it by default.
```suggestion
	enable: Boolean(String(process.env.ENABLE_APPLY_SNAPSHOT).toLowerCase() === 'true'), // Disabled by default
```

File: services/core/package.json [hunk 2]
```diff
@@ -47,6 +47,7 @@
     "method-override": "^3.0.0",
     "moment": "=2.24.0",
     "mysql": "^2.18.1",
+    "mysql-import": "^5.0.21",
```
Comment (sameersubudhi) [inline services/core/package.json:50]: Please remove the dependency since we're not using it anymore.

File: services/core/shared/downloadFile.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+const {
+	Logger,
+	Exceptions: { NotFoundException },
+} = require('lisk-service-framework');
+
+const logger = Logger();
+
+
+const downloadZip = (snapshotUrl, directoryPath) => new Promise((resolve, reject) => {
+	https.get(snapshotUrl, (response) => {
```
Comment (sameersubudhi) [inline services/core/shared/downloadFile.js:28]: Please consider adding unit tests for the utility.

```suggestion
const downloadAndExtractTarball = (fileUrl, directoryPath) => new Promise((resolve, reject) => {
	https.get(fileUrl, (response) => {
```

File: services/core/app.js [hunk 4]
```diff
@@ -42,24 +44,28 @@ const app = Microservice({
 	logger: loggerConf,
 });
 
-nodeStatus.waitForNode().then(async () => {
-	logger.info('Found a node, initiating Lisk Core...');
+snapshotUtils.initSnapshot()
+	.then(logger.info('Successfully downloaded and applied the snapshot'))
```
Comment (sameersubudhi) [inline services/core/app.js:48]:
```suggestion
	.then(() => {
		if (config.snapshot.enable)
			logger.info('Successfully downloaded and applied the snapshot')
	})
```

File: services/core/shared/downloadFile.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+			const errMessage = `Download failed with HTTP status code: ${response.statusCode} (${response.statusMessage})`;
+			logger.error(errMessage);
+			if (response.statusCode === 404) throw new NotFoundException(errMessage);
+			throw new Error(errMessage);
+		}
+	});
+});
+
+module.exports = {
+	downloadZip,
```
Comment (sameersubudhi) [inline services/core/shared/downloadFile.js:46]:
```suggestion
	downloadAndExtractTarball,
```

File: services/core/shared/core/compat/sdk_v5/blocksUtils.js [hunk 6]
```diff
@@ -14,23 +14,24 @@
...
-	Exceptions: { NotFoundException },
 } = require('lisk-service-framework');
 
 const {
 	exists,
 	mkdir,
 } = require('../../../fsUtils');
 
+const {
+	downloadZip,
```
Comment (sameersubudhi) [inline services/core/shared/core/compat/sdk_v5/blocksUtils.js:32]:
```suggestion
	downloadAndExtractTarball,
```

File: services/core/shared/core/compat/sdk_v5/blocksUtils.js [hunk 7]
```diff
@@ -92,23 +93,8 @@ const downloadGenesisBlock = async () => {
...
-					});
-				} else {
-					const errMessage = `Download failed with HTTP status code: ${response.statusCode} (${response.statusMessage})`;
-					logger.error(errMessage);
-					if (response.statusCode === 404) throw new NotFoundException(errMessage);
-					throw new Error(errMessage);
-				}
-			});
-		} else {
+		if (genesisBlockUrl.endsWith('.tar.gz')) downloadZip(genesisBlockUrl, directoryPath);
```
Comment (sameersubudhi) [inline services/core/shared/core/compat/sdk_v5/blocksUtils.js:96]:
```suggestion
		if (genesisBlockUrl.endsWith('.tar.gz')) downloadAndExtractTarball(genesisBlockUrl, directoryPath);
```

File: services/core/shared/core/compat/sdk_v5/blocksUtils.js [hunk 8]
```diff
@@ -92,23 +93,8 @@ const downloadGenesisBlock = async () => {
...
+		else {
 			request(genesisBlockUrl)
```
Comment (sameersubudhi) [inline services/core/shared/core/compat/sdk_v5/blocksUtils.js:98]: Please consider moving this utility as `downloadJSONFile` (or something similar) to `services/core/shared/downloadFile.js` as well.

File: services/core/shared/core/compat/sdk_v5/snapshotUtils.js [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+		logger.warn(stderr);
+		logger.info('SQL file(s) imported successfully');
+	} catch (error) {
+		logger.error(error);
+		throw error;
+	}
+};
+
+const initSnapshot = async () => {
+	if (config.snapshot.enable === false) {
```
Comment (sameersubudhi) [inline services/core/shared/core/compat/sdk_v5/snapshotUtils.js:85]: Keeping inline with the suggested change in config file.
```suggestion
	if (config.snapshot.enable !== true) {
```

File: services/core/shared/core/compat/sdk_v5/snapshotUtils.js [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+		logger.info('Index snapshot application has been disabled');
+		return;
+	}
+
+	const { data: { networkIdentifier } } = JSON.parse(await constantsCache.get('networkConstants'));
+
+	const [networkConfig] = config.netxworks.filter(c => networkIdentifier === c.identifier);
```
Comment (sameersubudhi) [inline services/core/shared/core/compat/sdk_v5/snapshotUtils.js:92]:
```suggestion
	const [networkConfig] = config.networks.filter(c => networkIdentifier === c.identifier);
```

File: services/core/shared/core/compat/sdk_v5/snapshotUtils.js [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+		snapshotFilePath = `./data/${networkIdentifier}/service-core-snapshot.sql`;
+	} else {
+		logger.info(`Network is neither defined in the config, nor in the environment variable (${networkIdentifier})`);
+		return;
+	}
+
+	if (!(await exists(snapshotFilePath))) {
+		await downloadSnapshot();
+	}
+	await applySnapshot();
```
Comment (sameersubudhi) [inline services/core/shared/core/compat/sdk_v5/snapshotUtils.js:104]:
```suggestion
	if (!(await exists(snapshotFilePath))) {
		await downloadSnapshot();
		await applySnapshot();
	}
```

File: services/core/shared/core/compat/sdk_v5/snapshotUtils.js [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ *
+ */
+const path = require('path');
+const util = require('util');
+const exec = util.promisify(require('child_process').exec);
```
Comment (sameersubudhi) [inline services/core/shared/core/compat/sdk_v5/snapshotUtils.js:18]: Please consider renaming the variable to something like `execInShell` and update wherever relevant.
```suggestion
const execInShell = util.promisify(require('child_process').exec);
```

**Review by sameersubudhi — CHANGES_REQUESTED** [review]
Kindly consider the following.

File: services/core/shared/core/compat/sdk_v5/snapshotUtils.js [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+};
+
+const resolveSnapshotRestoreCommand = async (connEndpoint) => {
+	await checkCommandAvailability();
+	const composeFile = config.snapshot.composeFilePath;
+	const dockerServiceName = config.snapshot.serviceName;
+
+	const [user, password] = connEndpoint.split('//')[1].split('@')[0].split(':');
+	const database = connEndpoint.split('@')[1].split('/')[1];
+	const mysqlSnapshotCommand = `mysql ${database} -u ${user} -p${password} < ${snapshotFilePath}`;
```
Comment (MichalTuleja) [inline services/core/shared/core/compat/sdk_v5/snapshotUtils.js:61]: Please add remote server support

File: services/core/shared/downloadFile.js [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+
+
+const downloadZip = (snapshotUrl, directoryPath) => new Promise((resolve, reject) => {
+	https.get(snapshotUrl, (response) => {
+		if (response.statusCode === 200) {
+			response.pipe(tar.extract({ cwd: directoryPath }));
+			response.on('error', async (err) => reject(err));
+			response.on('end', async () => {
+				logger.info('File downloaded successfully');
+				return setTimeout(resolve, 500);
```
Comment (MichalTuleja) [inline services/core/shared/downloadFile.js:34]: Would `setImmediate` work here?

**Review by MichalTuleja — CHANGES_REQUESTED** [review]

File: services/core/shared/downloadFile.js (hunk 14 above)
Comment (nagdahimanshu) [reply services/core/shared/downloadFile.js]: Yes, that would work. I updated the code

File: services/core/shared/downloadFile.js [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+
+const {
+	Logger,
+	Exceptions: { NotFoundException },
+	HTTP: { request },
+} = require('lisk-service-framework');
+
+const logger = Logger();
+
+const downloadAndExtractTarball = (url, directoryPath) => new Promise((resolve, reject) => {
```
Comment (sameersubudhi) [inline services/core/shared/downloadFile.js:28]: Can we also take this opportunity to implement a `downloadAndUnzipFile` method, for situations where we only have the snapshot zipped but not tarball (Eg: `*.sql.gz`)?

Ref: https://stackoverflow.com/questions/10308110/simplest-way-to-download-and-unzip-files-in-node-js-cross-platform. Please check the `zlib` module based solution.

Comment (nagdahimanshu) [reply services/core/shared/downloadFile.js:29]: I have implemented the required method `downloadAndUnzipFile` and updated the code.

File: services/core/package.json [hunk 16]
```diff
@@ -52,7 +52,8 @@
     "split": "^1.0.1",
     "stack-trace": "0.0.10",
     "string-similarity": "^4.0.4",
-    "tar": "^6.1.9"
+    "tar": "^6.1.9",
+    "zlib": "^1.0.5"
```
Comment (sameersubudhi) [inline services/core/package.json:56]: NodeJS by default has the `zlib` (https://nodejs.org/api/zlib.html#zlib) module available that also offers the `createUnzip` method. Kindly remove this dependency and re-test the implementation. I can confirm, the download utility works on my machine even without installing this dependency.

File: services/core/tests/functional/downloadFile.test.js [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+		const [{ snapshotUrl }] = config.networks.filter(network => network.name === 'testnet');
+		const filePath = `${directoryPath}/service-core-snapshot.sql`;
+		await downloadAndUnzipFile(snapshotUrl, filePath);
+		const isFileExists = !!(await fs.promises.stat(filePath).catch(() => null));
+		expect(isFileExists).toEqual(true);
+	});
+
+	it('downloadAndUnzipFile -> invalid url', async () => {
+		const url = 'https://downloads.lisk.com/lisk/testnet/service_core_invalid.sql';
+		const filePath = `${directoryPath}/service-core-snapshot.sql`;
```
Comment (sameersubudhi) [inline services/core/tests/functional/downloadFile.test.js:87]: Please use a different filename here than from the one above.
```suggestion
		const filePath = `${directoryPath}/service-core-snapshot-invalid.sql`;
```

**Review by sameersubudhi — CHANGES_REQUESTED** [review]

File: services/core/shared/core/compat/sdk_v5/snapshotUtils.js [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+ */
+const path = require('path');
+const util = require('util');
+const execInShell = util.promisify(require('child_process').exec);
+
+const {
+	CacheRedis,
+	Logger,
+	Exceptions: { NotFoundException },
+} = require('lisk-service-framework');
```
Comment (sameersubudhi) [inline services/core/shared/core/compat/sdk_v5/snapshotUtils.js:24]:
```suggestion
} = require('lisk-service-framework');

const {
	getNetworkConstants,
} = require('../common/constants');
```

File: services/core/shared/core/compat/sdk_v5/snapshotUtils.js [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+	}
+};
+
+const initSnapshot = async () => {
+	if (config.snapshot.enable !== true) {
+		logger.info('Index snapshot application has been disabled');
+		return;
+	}
+
+	const { data: { networkIdentifier } } = JSON.parse(await constantsCache.get('networkConstants'));
```
Comment (sameersubudhi) [inline services/core/shared/core/compat/sdk_v5/snapshotUtils.js:94]: Please make the following changes to handle the cold start scenario.
```suggestion
	const cachedNetworkConstants = await constantsCache.get('networkConstants');
	const {
		data: { networkIdentifier },
	} = cachedNetworkConstants ? JSON.parse(cachedNetworkConstants) : await getNetworkConstants();
```

**Review by sameersubudhi — APPROVED** [review]
Great job! Thank you for your contribution 🚀

**Review by MichalTuleja — APPROVED** [review]
🎉

**Pull request merged**